### PR TITLE
rgw: implement x-amz-replication-status for PENDING & COMPLETED

### PR DESCRIPTION
**Pull request by awojno-bloomberg**
Tracker: https://tracker.ceph.com/issues/64436

## Contribution Guidelines
- To sign and title your commits, please refer to [Submitting Patches to Ceph](https://github.com/ceph/ceph/blob/main/SubmittingPatches.rst).

- If you are submitting a fix for a stable branch (e.g. "quincy"), please refer to [Submitting Patches to Ceph - Backports](https://github.com/ceph/ceph/blob/master/SubmittingPatches-backports.rst) for the proper workflow.

- When filling out the below checklist, you may click boxes directly in the GitHub web UI.  When entering or editing the entire PR message in the GitHub web UI editor, you may also select a checklist item by adding an `x` between the brackets: `[x]`.  Spaces and capitalization matter when checking off items this way.

## Checklist
- Tracker (select at least one)
  - [x] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [ ] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [ ] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [x] No impact that needs to be tracked
- Documentation (select at least one)
  - [ ] Updates relevant documentation
  - [x] No doc update is appropriate
- Tests (select at least one)
  - [ ] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [x] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [ ] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`
- `jenkins test windows`
- `jenkins test rook e2e`
</details>

### CODE REVIEW

File: src/rgw/rgw_sal_filter.h
Comment (cbodley) [inline src/rgw/rgw_sal_filter.h:750]: return next->is_sync_completed()

File: src/rgw/rgw_op.cc
Comment (cbodley) [inline src/rgw/rgw_op.cc:925]: hmm there's a separate bug fix for `set_obj_attrs()` in https://github.com/ceph/ceph/pull/56275 that forces peers to re-replicate, but we don't want to do here. ideally our use here would preserve the existing mtime and avoid writing a bilog/datalog entry at all

cc @smanjara

File: src/rgw/rgw_op.cc
Comment (cbodley) [inline src/rgw/rgw_op.cc:921]: this `attrs` variable shadows the function parameter, please rename one

when we update RGW_ATTR_OBJ_REPLICATION_STATUS to "COMPLETED" here, does that updated value get sent in the x-amz-replication-status header?

File: src/rgw/rgw_sal_filter.h
Comment (awojno-bloomberg) [reply src/rgw/rgw_sal_filter.h]: updated

File: src/rgw/rgw_op.cc
Comment (awojno-bloomberg) [reply src/rgw/rgw_op.cc]: renamed the local `attrs` variable. Yes, this updated value is reflected in the header.

File: src/rgw/driver/rados/rgw_sal_rados.cc
Comment (awojno-bloomberg) [reply src/rgw/driver/rados/rgw_sal_rados.cc]: Changed styling and added a return code check

File: src/rgw/rgw_op.cc
Comment (awojno-bloomberg) [reply src/rgw/rgw_op.cc]: Updated

File: src/rgw/rgw_sync_policy.h
Comment (awojno-bloomberg) [reply src/rgw/rgw_sync_policy.h]: Added const modifier

File: src/rgw/rgw_op.cc
Comment (awojno-bloomberg) [reply src/rgw/rgw_op.cc]: If we don't update the object attrs then we would have to look up the bilog info every time, which doesn't seem ideal either. How should we proceed then? Open a tracker for the re-replication issue? Not sure if we would address that issue in this PR.

Comment (smanjara) [reply src/rgw/rgw_op.cc]: PR #56275  modifies the mtime to trigger replication. we don't want that here because you only want to update the `RGW_ATTR_OBJ_REPLICATION_STATUS`. you could probably just overwrite `mtime` that I am passing into `set_attrs()`  with the `obj_mtime` from `handle_replication_status_header` to preserve the existing mtime conditional upon the `FLAG_LOG_OP` that you are passing in. does that work?

Comment (cbodley) [reply src/rgw/rgw_op.cc]: https://github.com/ceph/ceph/pull/56275 should probably only update the mtime when FLAG_LOG_OP is set

Comment (cbodley) [reply src/rgw/rgw_op.cc]: what can we do to unblock both of these prs?

maybe the change to add flags to set_obj_attrs() should be split into a separate commit so it can be included in both prs for now

Comment (awojno-bloomberg) [reply src/rgw/rgw_op.cc]: If we merge this PR, couldn't https://github.com/ceph/ceph/pull/56275 just be rebased on top of these changes with some small changes to adjust for the flags. that PR is pretty small

Comment (smanjara) [reply src/rgw/rgw_op.cc]: if this pr isn't ready yet or still needs testing, then I would prefer a separate pr for setting the flag alone go in first so I can rebase on top of it.

Comment (cbodley) [reply src/rgw/rgw_op.cc]: this pr isn't ready to merge yet, while #56275 has already been through qa. maybe it's better to get that one merged first, and make the mtime change conditional on FLAG_LOG_OP in this pr. but ideally we could find a way to avoid blocking either pr on the merge of the other

what do you think @smanjara? what's left to do before merging #56275?

Comment (smanjara) [reply src/rgw/rgw_op.cc]: just the testcase which works with data checkpoint but not bucket checkpoint. the assert there makes sure that the sync was successful, so i don't think anything is blocking there.

Comment (awojno-bloomberg) [reply src/rgw/rgw_op.cc]: If we can fix the test in https://github.com/ceph/ceph/pull/56275 I can rebase and add that conditional logic. What is the timeline for that? These PRs have been open for 2+ months so it would be good to close them promptly

Comment (smanjara) [reply src/rgw/rgw_op.cc]: merged #56275. please go ahead and rebase @awojno-bloomberg

Comment (awojno-bloomberg) [reply src/rgw/rgw_op.cc]: @smanjara @cbodley I rebased and updated `rgw_rados.cc` to only update mtime if log_op is true. I am not aware of a scenario where we would want to update mtime but not add a replication log, but I can update the logic if there is a use case for that. Please take a look at the update when you can. Thanks

File: src/rgw/driver/rados/rgw_sal_rados.cc
Comment (smanjara) [inline src/rgw/driver/rados/rgw_sal_rados.cc:2325]: here in line [2330](https://github.com/ceph/ceph/pull/57060/files#diff-8d5ab77cf6b24e95c2c7b352c52d1e254781f6c9e7fb3fd148bddc8a3f4321e7R2330) we adjust mtime to trigger replication on a modified object. now, since we don't want that to happen in case of `handle_replication_status_header()`, you are doing the right thing by passing in value 0 for `flags`. we just need to change the code here to modify mtime only based on the condition that `FLAG_LOG_OP` is set and not otherwise.

File: src/rgw/rgw_sal_dbstore.cc
Comment (smanjara) [inline src/rgw/rgw_sal_dbstore.cc:564]: rgw::sal::FLAG_LOG_OP instead of 0?

File: src/rgw/rgw_sal_dbstore.cc
Comment (smanjara) [inline src/rgw/rgw_sal_dbstore.cc:554]: rgw::sal::FLAG_LOG_OP instead of 0?

File: src/rgw/rgw_sal_dbstore.cc
Comment (awojno-bloomberg) [reply src/rgw/rgw_sal_dbstore.cc]: Updated to `rgw::sal::FLAG_LOG_OP`

File: src/rgw/rgw_sal_dbstore.cc
Comment (awojno-bloomberg) [reply src/rgw/rgw_sal_dbstore.cc]: Updated to `rgw::sal::FLAG_LOG_OP`

File: src/rgw/driver/rados/rgw_bucket_sync.h
Comment (awojno-bloomberg) [reply src/rgw/driver/rados/rgw_bucket_sync.h]: Added `const` modifier

File: src/rgw/rgw_op.cc
Comment (awojno-bloomberg) [reply src/rgw/rgw_op.cc]: Updated to use move assignment

File: src/rgw/driver/rados/rgw_sal_rados.cc
Comment (awojno-bloomberg) [reply src/rgw/driver/rados/rgw_sal_rados.cc:2325]: Updated logic to only increment mtime when `FLAG_LOG_OP` is set. Please review updated logic

File: src/rgw/driver/rados/rgw_rados.cc
Comment (awojno-bloomberg) [reply src/rgw/driver/rados/rgw_rados.cc]: Removed this logic and added conditional logic in `rgw_sal_rados.cc` to increment mtime based on `FLAG_LOG_OP`

File: src/rgw/rgw_op.cc
Comment (cbodley) [inline src/rgw/rgw_op.cc:4487]: it looks like `obj_tags` can be null here. seeing crashes in qa:
```
2024-06-21T02:14:03.883 INFO:tasks.rgw.client.0.smithi175.stdout:/usr/include/c++/11/bits/unique_ptr.h:407: typename std::add_lvalue_reference<_Tp>::type std::unique_ptr<_Tp, _Dp>::operator*() const [with _Tp = RGWObjTags; _Dp = std::default_delete<RGWObjTags>; typename std::add_lvalue_reference<_Tp>::type = RGWObjTags&]: Assertion 'get() != pointer()' failed.
2024-06-21T02:14:03.884 INFO:tasks.rgw.client.0.smithi175.stdout:*** Caught signal (Aborted) **
2024-06-21T02:14:03.884 INFO:tasks.rgw.client.0.smithi175.stdout: in thread 7faf3dc40640 thread_name:io_context_pool
2024-06-21T02:14:03.884 INFO:tasks.rgw.client.0.smithi175.stdout: ceph version 19.0.0-4398-g348e6f9d (348e6f9d58b67fa11bf46aaeb2d3fd0b12b25a07) squid (dev)
2024-06-21T02:14:03.884 INFO:tasks.rgw.client.0.smithi175.stdout: 1: /lib64/libc.so.6(+0x3e6f0) [0x7faf9d83e6f0]
2024-06-21T02:14:03.885 INFO:tasks.rgw.client.0.smithi175.stdout: 2: /lib64/libc.so.6(+0x8b94c) [0x7faf9d88b94c]
2024-06-21T02:14:03.885 INFO:tasks.rgw.client.0.smithi175.stdout: 3: raise()
2024-06-21T02:14:03.885 INFO:tasks.rgw.client.0.smithi175.stdout: 4: abort()
2024-06-21T02:14:03.885 INFO:tasks.rgw.client.0.smithi175.stdout: 5: radosgw(+0x400cc8) [0x55ec7da07cc8]
2024-06-21T02:14:03.885 INFO:tasks.rgw.client.0.smithi175.stdout: 6: (RGWPutObj::execute(optional_yield)+0x4647) [0x55ec7dc8c177]
2024-06-21T02:14:03.885 INFO:tasks.rgw.client.0.smithi175.stdout: 7: (rgw_process_authenticated(RGWHandler_REST*, RGWOp*&, RGWRequest*, req_state*, optional_yield, rgw::sal::Driver*, bool)+0xaaa) [0x55ec7db1faaa]
2024-06-21T02:14:03.885 INFO:tasks.rgw.client.0.smithi175.stdout: 8: (process_request(RGWProcessEnv const&, RGWRequest*, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, RGWRestfulIO*, optional_yield, rgw::dmclock::Scheduler*, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >*, std::chrono::duration<unsigned long, std::ratio<1l, 1000000000l> >*, int*)+0xfed) [0x55ec7db212ad]
2024-06-21T02:14:03.885 INFO:tasks.rgw.client.0.smithi175.stdout: 9: radosgw(+0xe3a11f) [0x55ec7e44111f]
2024-06-21T02:14:03.885 INFO:tasks.rgw.client.0.smithi175.stdout: 10: radosgw(+0x477816) [0x55ec7da7e816]
2024-06-21T02:14:03.885 INFO:tasks.rgw.client.0.smithi175.stdout: 11: radosgw(+0x459dc4) [0x55ec7da60dc4]
2024-06-21T02:14:03.885 INFO:tasks.rgw.client.0.smithi175.stdout: 12: make_fcontext()
```

**Comment by cbodley** [comment]
failed qa in https://pulpito.ceph.com/cbodley-2024-06-20_13:04:58-rgw-wip-cbodley-testing-distro-default-smithi/

File: src/rgw/rgw_op.cc
Comment (awojno-bloomberg) [reply src/rgw/rgw_op.cc]: Updated to use the safer `.get()` and edited related methods. Please re-test when possible

File: src/rgw/driver/rados/rgw_bucket_sync.cc
Comment (cbodley) [inline src/rgw/driver/rados/rgw_bucket_sync.cc:993]: this `tags == nullptr` part seems backwards to me. if a [bucket replication policy](https://docs.aws.amazon.com/AmazonS3/latest/API/API_PutBucketReplication.html) contains a tag-based filter like:
```
<Filter>
    <Tag>
        <Key>string</Key>
        <Value>string</Value>
    </Tag>
</Filter>
```
then we should only sync objects that have a matching tag. if an object has no tags, we shouldn't return true here

it seems like `bucket_exports_object()` should require a `tags` argument (take as a const ref), and callers should pass an empty RGWObjTags instance rather than nullptr

that requires changes to RGWPutObj to avoid passing in nullptr. i prepared https://github.com/ceph/ceph/pull/58291 as an example - if that's useful, feel free to include that commit here

**Review by cbodley — APPROVED** [review]
thanks, looks good

**Comment by cbodley** [comment]
passed qa in https://pulpito.ceph.com/cbodley-2024-07-03_01:53:26-rgw-wip-cbodley-testing-distro-default-smithi/

> 2024-07-03T02:41:10.315 INFO:tasks.rgw_multisite_tests:rgw_multi.tests.test_replication_status ... ok

**Comment by jmundack** [comment]
thanks @cbodley !

File: src/rgw/rgw_op.cc
Comment (cbodley) [inline src/rgw/rgw_op.cc:927]: cc @smanjara @thotz 

i've been seeing multisite failures recently, and this seems to be the culprit. the underlying bug is https://tracker.ceph.com/issues/64173, whose fix in https://github.com/ceph/ceph/pull/54393 still hasn't merged

to summarize, this call to `set_obj_attrs()` does a bucket index transaction, but passes empty values for the owner, etag, storage class, etc. this causes ListObjects to return empty values for Owner and ETag, leading to test failures like this:
```
======================================================================
FAIL: rgw_multi.tests.test_bucket_reshard_incremental
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/teuthworker/src/git.ceph.com_teuthology_454a46106b61e201598ae5e1c392af2328539bfe/virtualenv/lib/python3.10/site-packages/nose/case.py", line 170, in runTest
    self.test(*self.arg)
  File "/home/teuthworker/src/git.ceph.com_ceph-c_532ed64a904276d871b91839ffec7885dab6b986/qa/../src/test/rgw/rgw_multi/tests.py", line 1692, in test_bucket_reshard_incremental
    zonegroup_bucket_checkpoint(zonegroup_conns, bucket.name)
  File "/home/teuthworker/src/git.ceph.com_ceph-c_532ed64a904276d871b91839ffec7885dab6b986/qa/../src/test/rgw/rgw_multi/tests.py", line 401, in zonegroup_bucket_checkpoint
    target_conn.check_bucket_eq(source_conn, bucket_name)
  File "/home/teuthworker/src/git.ceph.com_ceph-c_532ed64a904276d871b91839ffec7885dab6b986/qa/tasks/rgw_multi/zone_rados.py", line 84, in check_bucket_eq
    check_object_eq(k1, k2)
  File "/home/teuthworker/src/git.ceph.com_ceph-c_532ed64a904276d871b91839ffec7885dab6b986/qa/tasks/rgw_multi/zone_rados.py", line 34, in check_object_eq
    eq(k1.owner.id, k2.owner.id)
AssertionError: 'RGW11111111111111111' != ''
```
when comparing the output of `radosgw-admin bucket list` and `object stat`, i see that objects with "user.rgw.amz-replication-status": "COMPLETE" have empty values in the index:
```
        "meta": {
            "category": 1,
            "size": 3,
            "mtime": "2024-07-11T02:44:52.655546Z",
            "etag": "",
            "storage_class": "",
            "owner": "",
            "owner_display_name": "",
            "content_type": "",
```
while the entries are still intact for objects with "user.rgw.amz-replication-status": "PENDING":
```
        "meta": {
            "category": 1,
            "size": 3,
            "mtime": "2024-07-11T02:44:52.610456Z",
            "etag": "acbd18db4cc2f85cedef654fccc4a4d8",
            "storage_class": "",
            "owner": "RGW11111111111111111",
            "owner_display_name": "",
            "content_type": "application/octet-stream",
```

Comment (awojno-bloomberg) [reply src/rgw/rgw_op.cc:927]: Can we work to merge the PR that fixes the underlying bug? Anything I can do to help out?

Comment (cbodley) [reply src/rgw/rgw_op.cc:927]: thanks @awojno-bloomberg. i'm trying to get this unblocked in https://github.com/ceph/ceph/pull/54393#pullrequestreview-2172100453. you might help by testing that out to verify that it fixes the transition to COMPLETED?

Comment (cbodley) [reply src/rgw/rgw_op.cc:927]: fix merged in https://github.com/ceph/ceph/pull/58649

**Pull request merged**
